gradle-plugin/src/main/java/com/squareup/hephaestus/plugin/HephaestusPlugin.kt

```diff
@@ -58,101 +78,129 @@ open class HephaestusPlugin : Plugin<Project> {
-          }
-
-          val configurationNameForIncrementalChecks = when {
-            // Disable incremental compilation, if the compiler plugin dependency isn't up-to-date.
-            // This will trigger a full compilation of a module using Hephaestus even though its
-            // source files might not have changed. This workaround is necessary, otherwise
-            // incremental builds are broken. See https://youtrack.jetbrains.com/issue/KT-38570
-            !isStubGeneratingTask -> PLUGIN_CLASSPATH_CONFIGURATION_NAME
-
-            // For Java / Kotlin projects it doesn't change.
-            project.isKotlinJvmProject -> "compileClasspath"
-
-            // Disable incremental compilation for the stub generating task, if any of the
-            // dependencies in the compile classpath have changed. This will make sure that we pick
-            // up any change from a dependency when merging all the classes. Without this workaround
-            // we could make changes in any library, but these changes wouldn't be contributed to the
-            // Dagger graph, because incremental compilation tricked us.
-            else -> {
-              val variant = project.androidVariants()
-                  .findVariantForCompileTask(compileTask)
-
-              "${variant.name}CompileClasspath"
-            }
+        .withType(KaptGenerateStubsTask::class.java)
+        .configureEach { stubsTask ->
+          // Disable incremental compilation for the stub generating task. If any of the
+          // dependencies in the compile classpath have changed, then this we need to trigger the
+          // compiler plugin. This will make sure that we pick up any change from a dependency
+          // when merging all the classes. Without this workaround we could make changes in any
+          // library, but these changes wouldn't be contributed to the Dagger graph, because
+          // incremental compilation tricked us.
+          stubsTask.doFirst {
+            stubsTask.incremental = false
+            stubsTask.logger.info(
+                "Hephaestus: Incremental compilation enabled: ${stubsTask.incremental} (stub)"
+            )
           }
+        }
 
-          val checkIfCompilerPluginChangedTaskProvider = project.tasks
-              .register(
-                  compileTask.name + "CheckIncrementalCompilationHephaestus",
-                  DisableIncrementalCompilationTask::class.java
-              ) {
-                it.compileTask = compileTask
-                it.configurationName = configurationNameForIncrementalChecks
-              }
+    // Use this signal to share state between DisableIncrementalCompilationTask and the Kotlin
+    // compile task. If the plugin classpath changed, then DisableIncrementalCompilationTask sets
+    // the signal to false.
+    val incrementalSignal = project.gradle.sharedServices
+        .registerIfAbsent("incrementalSignal${project.path}", IncrementalSignal::class.java) { }
```

Comment:
Would it make sense to register just a single shared service, and have that maintain a `Map<String, Boolean?>` of incremental signals, instead of one service per project?

Reply:
I honestly have no idea. I would hope that this would have only a tiny impact and not be measurable at all. I'll change it, because it somehow sounds safer.